pkg/server/singleprocess/service.go

```diff
@@ -0,0 +1,359 @@
+	decodeId func(encodedId string) (id string, err error)
+
+	logStreamProvider logstream.Provider
+
+	// urlConfig is not nil if the URL service is enabled. This is guaranteed
+	// to have the configs set.
+	urlConfig    *serverconfig.URL
+	urlClientMu  sync.Mutex
+	urlClientVal wphznpb.WaypointHznClient
+
+	// urlCEB has the configuration for the entrypoint. If this is nil,
+	// then the configuration is not ready. The urlCEBWatchCh can be used
+	// to watch for changes. All fields protected with urlCEBMu.
+	urlCEBMu      sync.RWMutex
+	urlCEB        *pb.EntrypointConfig_URLService
+	urlCEBWatchCh chan struct{}
+
+	// bgCtx is used for background tasks within the service. This is
+	// cancelled when Close is called.
+	bgCtx       context.Context
+	bgCtxCancel context.CancelFunc
+
+	// bgWg is incremented for every background goroutine that the
+	// service starts up. When Close is called, we wait on this to ensure
+	// that we fully shut down before returning.
+	bgWg sync.WaitGroup
+
+	// superuser is true if all API actions should act as if a superuser
+	// made them. This is used for local mode only.
+	superuser bool
+
+	// oidcCache is the cache for OIDC providers.
+	oidcCache *wpoidc.ProviderCache
+}
+
+// New returns a Waypoint server implementation that uses BotlDB plus
+// in-memory locks to operate safely.
+func New(opts ...Option) (pb.WaypointServer, error) {
+	var s Service
+
+	// Set default config values
+	cfg := config{
+		idEncoder: func(ctx context.Context, id string) (encodedId string, err error) {
+			return id, nil
+		},
+		idDecoder: func(encodedId string) (id string, err error) {
+			return encodedId, nil
+		},
+		logStreamProvider: &singleProcessLogStreamProvider{},
+	}
+
+	for _, opt := range opts {
+		if err := opt(&s, &cfg); err != nil {
+			return nil, err
+		}
+	}
+
+	s.encodeId = cfg.idEncoder
+	s.decodeId = cfg.idDecoder
+
+	if !cfg.oidcDisabled {
+		s.oidcCache = wpoidc.NewProviderCache()
+	}
+
+	if cfg.logStreamProvider != nil {
+		s.logStreamProvider = cfg.logStreamProvider
+	}
+
+	log := cfg.log
+	if log == nil {
+		log = hclog.L()
+	}
+
+	s.state = cfg.stateProvider
+
+	// If we don't have a server ID, set that.
+
+	// TODO(izaak): the serverstate interface doesn't currently support
+	// the ServerId methods, but I think it probably needs to.
+	state := s.state(context.Background())
+	if boltdbstate, ok := state.(*boltdbstate.State); ok {
+		id, err := boltdbstate.ServerIdGet()
+		if err != nil {
+			return nil, err
+		}
+		if id == "" {
+			id, err = server.Id()
+			if err != nil {
+				return nil, err
+			}
+
+			if err := boltdbstate.ServerIdSet(id); err != nil {
+				return nil, err
+			}
+		}
+		s.id = id
+
+		// If we haven't initialized our server config before, do that once.
+		conf, err := state.ServerConfigGet()
+		if err != nil {
+			return nil, err
+		}
+		if conf.Cookie == "" {
+			if err := state.ServerConfigSet(conf); err != nil {
+				return nil, err
+			}
+		}
+	}
```

Comment:
Not sure if this code was copied in the move but we should probably handle the not `ok` case here? Return an `codes.FailedPrecondition` error?

Reply:
This was a bit of a hack - inside this `if` is where I was hiding the logic that was only relevant to the singleprocess server, but didn't have good configuration flags for.

After thinking about it a bit, I managed to find better homes for this logic.